Refactor the `Account` state

PhilippGackstatter opened the pull request:
# Description of change

This PR implements part of what was discussed in [#433](https://github.com/iotaledger/identity.rs/issues/433), namely the refactoring of the state representation in the `Account`.

The main changes:

- Remove the event & commit system of the account and use a dual-document state. One document, the current on-tangle representation, is cached locally in storage, while all updates are applied to an in-memory document that is part of the account's state.
- Remove the parts of the `IdentityState` that were essentially duplicates of the `IotaDocument`, e.g. `TinyMethod`, `TinyService`, etc. The state type now holds a document directly and only adds some state on top.
- Remove diff and int generation from the state, and replace it with a single generation field. These generations are used for detecting whether an identity is fresh (generation == 0) and used in the location of keys to differentiate between keys with the same fragment, that were created in different generations.
  - The current solution is temporary until https://github.com/iotaledger/stronghold.rs/issues/283 lands, which will allow us to implement key locations based on the public key and the fragment, and remove all extra state such as generations.
  - Once the above is implemented, the `IdentityState` will be removed completely and the account can hold the `IotaDocument` directly.
  - In anticipation of the complete removal of `IdentityState`, the chain state was factored out of that type into a dedicated `ChainState` type. This keeps track of the previous integration and diff chain message id, which is required state for implementing automatic int or diff publishing.
  - The check for whether an identity is new is based on whether the previous integration message id is `MessageId::null()`, since the generation is anticipated to be removed in future updates.
- Introduce `MethodRelationship` next to `MethodScope`, in order to disallow attaching relationships to a method with `MethodScope::VerificationMethod`.

This PR does not address:

- Moving the existence and duplication checks for insertion and removal of methods to the `IotaDocument`.
- Purging/Deleting keys in storage, this will be addressed with the key location refactor, although deleting keys as soon as they are no longer needed is slightly risky (if an update isn't included in the tangle) and hasn't been addressed in the account so far, so this might need some discussion.
- Saving every update as a diff to storage. The state is currently only saved to storage on identity creation and after a successful publication. Dropping the account, with several unpublished updates in-memory, would get rid of these updates. A future PR will implement storing the diffs of each update in storage for later publication, or possible re-application on a changed base document (e.g. one that was resolved/synced from the tangle). One of the light blockers on including it in this PR, is that diffing metadata is currently hard, because `identity_iota::did::doc::properties::Properties` does not implement `Diff`. #486 will presumably make diffing metadata easier than it currently is, which should make it easier to implement this feature.

Since this is a large PR, and anyone would like to focus on a specific area of it, my suggestion would be to carefully review the publication logic around whether to publish as an integration or diff update, and whether the `previous_message_id` is always set correctly.

Finally, while this PR does not address every part of the planned refactor, it should be good enough to get started with the low-level access API implementation and the synchronization.

## Links to any relevant issues

#433

## Type of change
Add an `x` to the boxes that are relevant to your changes.

- [ ] Bug fix (a non-breaking change which fixes an issue)
- [ ] Enhancement (a non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Fix

## How the change has been tested

Several tests were added or extended, namely the updates test, `Publish` tests, and adding or detaching relationships on an `IotaDocument`.

## Change checklist
Add an `x` to the boxes that are relevant to your changes.

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

cycraig commented on identity-account/src/updates/update.rs:
```diff
@@ -0,0 +1,465 @@
+          UpdateError::MethodNotFound
+        );
+
+        // TODO: Do we have to ? here?
+        let method_url: IotaDIDUrl = did.to_url().join(fragment.identifier())?;
+        let core_method_url: CoreDIDUrl = CoreDIDUrl::from(method_url.clone());
```
I think so, yes. The fragment is user-input and could contain invalid or reserved characters. `Fragment` is separate from `DIDUrl` (annoyingly) and doesn't validate the contents.

cycraig commented on identity-account/src/updates/update.rs:
```diff
@@ -0,0 +1,465 @@
+pub const UPDATE_METHOD_TYPES: &[MethodType] = &[MethodType::Ed25519VerificationKey2018];
+pub const DEFAULT_UPDATE_METHOD_PREFIX: &str = "sign-";
+
+fn key_to_method(type_: KeyType) -> MethodType {
+  match type_ {
+    KeyType::Ed25519 => MethodType::Ed25519VerificationKey2018,
+  }
+}
```
I think this is only used in one place now, maybe remove the function and inline the code?

cycraig commented on identity-account/src/updates/update.rs:
```diff
@@ -0,0 +1,465 @@
+
+  let method_fragment = location.fragment().to_owned();
+
+  // TODO: Can we unwrap/expect?
+  let method: IotaVerificationMethod =
+    IotaVerificationMethod::from_did(did, setup.key_type, &public, method_fragment.name())?;
+
+  // TODO: Can we unwrap/expect?
+  let document = IotaDocument::from_verification_method(method)?;
```
There may be unforeseen errors in the future, safer to use `?` or map it to some other error if desired.

cycraig commented on identity-account/src/updates/update.rs:
```diff
@@ -0,0 +1,465 @@
+          UpdateError::InvalidMethodFragment("cannot remove last signing method")
+        );
+
+        state.document_mut().remove_method(method_url).unwrap();
```
Use `?` instead of `unwrap`?

cycraig commented on identity-account/src/updates/update.rs:
```diff
@@ -0,0 +1,465 @@
+          UpdateError::MethodNotFound
+        );
+
+        // The verification method must not be embedded.
+        ensure!(
+          !state
+            .document()
+            .as_document()
+            .verification_relationships()
+            .any(|method_ref| match method_ref {
+              MethodRef::Embed(method) => method.id().fragment() == method_url.fragment(),
+              MethodRef::Refer(_) => false,
+            }),
+          UpdateError::InvalidTargetMethod
+        );
```
I think this was discussed before but is it not possible to attach methods to more than one verification relationship even if it's embedded already?

PhilippGackstatter replied:
I thought with the agreement on introducing `MethodRelationship` the consensus was that attaching relationships on embedded methods is an error. Otherwise, that method would have to be moved to the verification method set and a reference would have to be inserted in the place where it was previously embedded. That's not terribly complicated, but I don't know if it's expected behavior by users?

cycraig replied:
I'm unsure. I was going by the behaviour of `IotaDocument::attach_method_relationship` which only checks that the method is part of the document (as any relationship or just a verification method) then adds another reference to it, unless I'm reading it wrong? Basically the low-level API now allows attaching multiple relationships to embedded methods, but not the `Account`. (code below)

I'm still unsure if the DID specification restricts that a method with multiple relationships MUST be a `verificationMethod` and not embedded. https://www.w3.org/TR/did-core/#verification-relationships

```rust
// Attaches the given relationship to the given method, if the method exists.
pub fn attach_method_relationship(&mut self, did_url: IotaDIDUrl, relationship: MethodRelationship) -> Result<bool> {
  // Ensure the method exists
  self
    .document
    .resolve_method(did_url.url()) <============= HERE: will return any method, even an embedded one
    .ok_or(identity_did::Error::QueryMethodNotFound)?;
  
  let core_did_url: CoreDIDUrl = CoreDIDUrl::from(did_url);
  let method_ref = MethodRef::Refer(core_did_url);
  
  let result = match relationship {
    MethodRelationship::Authentication => self.document.authentication_mut().append(method_ref),
    MethodRelationship::AssertionMethod => self.document.assertion_method_mut().append(method_ref),
    MethodRelationship::KeyAgreement => self.document.key_agreement_mut().append(method_ref),
    MethodRelationship::CapabilityDelegation => self.document.capability_delegation_mut().append(method_ref),
    MethodRelationship::CapabilityInvocation => self.document.capability_invocation_mut().append(method_ref),
  };

  Ok(result)
}
```

cycraig replied:
Actually, I'm not quite right on the behaviour of `resolve_method`. It depends on the order that the relationships are added. 

E.g. 
1. if a method is embedded as an `Authentication` relationship then `attach`ed as an `AssertionMethod` reference, `resolve_method` will return the embedded method.
2. if a method is embedded as a `CapabilityInvocation` relationship then `attach`ed as an `Authentication` reference, `resolve_method` will return `None`.

This is due to the following code, note the match of `Method::Refer` only looks at `verificationMethod`.

```rust
  fn resolve_method_inner(&self, query: MethodQuery<'_>) -> Option<&VerificationMethod<U>> {
    let mut method: Option<&MethodRef<U>> = None;

    if method.is_none() {
      method = self.authentication.query(query.clone());
    }

    if method.is_none() {
      method = self.assertion_method.query(query.clone());
    }

    if method.is_none() {
      method = self.key_agreement.query(query.clone());
    }

    if method.is_none() {
      method = self.capability_delegation.query(query.clone());
    }

    if method.is_none() {
      method = self.capability_invocation.query(query.clone());
    }

    match method {
      Some(MethodRef::Embed(method)) => Some(method),
      Some(MethodRef::Refer(did)) => self.verification_method.query(&did.to_string()),
      None => self.verification_method.query(query),
    }
  }
```

**Conclusion:** based on this code, you are correct that we should *not* allow references to methods embedded as relationships. My suggestion would be to update `attach_method_relationship` to use `resolve_method_with_scope(method, MethodScope::VerificationMethod)` and return an error like `MethodEmbeddedOrNotFound`. A few new tests might be needed.

The alternative would be to update `resolve_method_inner` to resolve embedded references properly. Let me know if any of my assumptions here are incorrect.

PhilippGackstatter replied:
Added the `resolve_method_with_scope` check and kept the original error. Will make the errors better in the future, when the `UpdateError::InvalidTargetEmbeddedMethod` moves to the lower-level API.

cycraig commented on identity-account/src/updates/update.rs:
```diff
@@ -0,0 +1,465 @@
+      } => {
+        let fragment: Fragment = Fragment::new(fragment);
+
+        // TODO: Do we have to ? here?
+        let method_url: IotaDIDUrl = did.to_url().join(fragment.identifier())?;
```
Yes, any user input could be an invalid fragment.

cycraig commented on identity-account/src/updates/update.rs:
```diff
@@ -0,0 +1,465 @@
+/// # Parameters
+/// - `relationships`: the relationships to add, defaults to an empty [`Vec`].
+/// - `fragment`: the identifier of the method in the document, required.
+AttachMethod {
+  @required fragment String,
+  @default relationships Vec<MethodRelationship>,
+});
```
Discussion: should `AttachMethod` and `DetachMethod` take in a single relationship instead of a `Vec`? I know the PR didn't change this, just wondering since creating a `Vec` isn't very ergonomic, even more so in bindings.

PhilippGackstatter replied:
It would make error handling easier, if we want to return an error if a relationship wasn't found/already existed. If 4 relationships were added successfully, but the fifth one was already there, then what do you return? On the other hand, this is an argument to not return an error for these methods, because they're idempotent.

Also, the ergonomics are quite nice, since you don't have to pass a vector but can just call `relationship` multiple times to add multiple relationships, e.g.:

```rust
  account
    .update_identity()
    .attach_method()
    .fragment("my-next-key")
    .relationship(MethodRelationship::CapabilityDelegation)
    .relationship(MethodRelationship::CapabilityInvocation)
    .apply()
    .await?;
```

As long as the builder works nicely in other languages, that should translate nicely. If that's not idiomatic, then for example in JS or Python, passing a list seems nice, too. E.g.:

```python
await account.attach_method("my-next-key", [MethodRelationship.CapabilityDelegation, MethodRelationShip.CapabilityInvocation])
```

Overall, I think it's fine.

cycraig replied:
> Also, the ergonomics are quite nice, since you don't have to pass a vector but can just call relationship multiple times to add multiple relationships, e.g.:

Good point, I didn't catch that!

> On the other hand, this is an argument to not return an error for these methods, because they're idempotent.

This change would be for if those methods were not idempotent, as long as they are (and given the builder ergonomics) this is absolutely fine. Thank you.

cycraig commented on identity-account/src/updates/update.rs:
```diff
@@ -0,0 +1,465 @@
+        );
+
+        for relationship in relationships {
+          // We ignore the boolean result: if the relationship already existed, that's fine.
+          state
+            .document_mut()
+            .attach_method_relationship(method_url.clone(), relationship)
+            .map_err(|_| UpdateError::MethodNotFound)?;
```
Nit: explicitly ignore the result.

```suggestion
          // Ignore the boolean result: if the relationship already existed, that's fine.
          let _ = state
            .document_mut()
            .attach_method_relationship(method_url.clone(), relationship)
            .map_err(|_| UpdateError::MethodNotFound)?;
```

cycraig commented on identity-account/src/updates/update.rs:
```diff
@@ -0,0 +1,465 @@
+          UpdateError::MethodNotFound
+        );
+
+        // TODO: Do we have to ? here?
+        let method_url: IotaDIDUrl = did.to_url().join(fragment.identifier())?;
+        let core_method_url: CoreDIDUrl = CoreDIDUrl::from(method_url.clone());
```
Yes, any user input could be an invalid fragment.

cycraig commented:
> Saving every update as a diff to storage. The state is currently only saved to storage on identity creation and after a successful publication. Dropping the account, with several unpublished updates in-memory, would get rid of these updates. A future PR will implement storing the diffs of each update in storage for later publication, or possible re-application on a changed base document (e.g. one that was resolved/synced from the tangle).

Can't we store the altered document instead of the diffs? Applying/re-applying updates to a new base document (synchronising) could be achieved by first calculating the diff between the old and new document, fetching the latest base document (replacing old) then attempting to apply the diff to it? 
I'm just looking for a more straightforward solution than storing diffs.

cycraig commented on identity-account/src/account/account.rs:
```diff
@@ -146,25 +148,36 @@ impl Account {
+  // Identity
+  // ===========================================================================
+
+  /// Resolves the DID Document associated with this `Account` from the Tangle.
+  pub async fn resolve_identity(&self) -> Result<IotaDocument> {
+    // Fetch the DID Document from the Tangle
+    self.client_map.resolve(self.did()).await.map_err(Into::into)
```
Nit: redundant comment.
```suggestion
  /// Resolves the DID Document associated with this `Account` from the Tangle.
  pub async fn resolve_identity(&self) -> Result<IotaDocument> {
    self.client_map.resolve(self.did()).await.map_err(Into::into)
```

cycraig commented on identity-account/src/identity/chain_state.rs:
```diff
@@ -0,0 +1,63 @@
+  }
+
+  /// Returns the previous diff message id, or the previous integration message id.
+  pub fn diff_message_id(&self) -> &MessageId {
```
Nit: should this be `previous_diff_message_id` for consistency?

cycraig commented on identity-account/src/identity/chain_state.rs:
```diff
@@ -0,0 +1,63 @@
+    self.previous_diff_message_id = MessageId::null();
+  }
+
+  /// Sets the current Tangle diff message id of the identity.
+  pub fn set_diff_message_id(&mut self, message: MessageId) {
+    self.previous_diff_message_id = message;
+  }
```
Nit: should this be `set_previous_diff_message_id` for consistency?

cycraig commented on identity-account/src/tests/updates.rs:
```diff
@@ -0,0 +1,656 @@
+    Error::UpdateError(UpdateError::DuplicateKeyLocation(_)),
+  ));
+
+  // Fake publishing by incrementing the generation.
+  account.state_mut_unchecked().increment_generation().unwrap();
```
Can we instead set `testmode` to true on the `Account` and call `publish_updates`? Not critical if generations are being removed.

cycraig commented on identity-account/src/identity/chain_state.rs:
```diff
@@ -0,0 +1,63 @@
+
+/// Holds the last published message ids of the integration and diff chains.
+#[derive(Clone, Debug, PartialEq, Deserialize, Serialize)]
+pub struct ChainState {
+  #[serde(default = "MessageId::null", skip_serializing_if = "MessageId::is_null")]
+  previous_integration_message_id: MessageId,
+  #[serde(default = "MessageId::null", skip_serializing_if = "MessageId::is_null")]
+  previous_diff_message_id: MessageId,
+}
```
Nit: naming these just `integration_message_id` and `diff_message_id` might be fine. Very optional, not strongly opposed either way.

PhilippGackstatter replied:
I would like something that indicates that these are the ids of the messages that we have to attach the next update to, in order for it to be valid. The current naming is at least in line with the `previous_message_id` of the document, while I think a `last_` prefix would also work.

cycraig replied:
I like `last_`! Less conflict with existing terms I think.

cycraig commented on identity-account/src/identity/chain_state.rs:
```diff
@@ -0,0 +1,63 @@
+
+/// Holds the last published message ids of the integration and diff chains.
+#[derive(Clone, Debug, PartialEq, Deserialize, Serialize)]
+pub struct ChainState {
+  #[serde(default = "MessageId::null", skip_serializing_if = "MessageId::is_null")]
+  previous_integration_message_id: MessageId,
```
I wonder if the `ChainState` could be removed if e.g. `ResolutionMetadata` contains these `message_ids` and is stored with the old state. That would assume the `Account` resolves the document from the Tangle after every update though...

No changes requested, just wondering about the metadata refactor.

PhilippGackstatter replied:
That would be ideal! I think that should actually work because the last published document (= latest document on tangle) is always stored in storage. Then the two cases should work:

1. In a non-distributed setup, only one device/machine/program updates a given identity, and so taking the message ids from the last published document's metadata would work.
2. In a distributed setup, a valid update is only possible if the latest state has been synchronized before, which is why we require developers of such setups to call sync before publishing an update. Then we'd also have the latest message ids locally. If someone fails to synchronize before, then no matter if we hold the state in `ChainState`, or if we take it from the metadata, it will be outdated and the update will be ignored.

cycraig replied:
That makes sense. I'm just thinking about other fields that depend on resolution (like `version`) that cannot/should not be updated locally. Anyway, I'll keep this in mind during implementation.

cycraig commented on identity-account/src/identity/chain_state.rs:
```diff
@@ -0,0 +1,63 @@
+    &self.previous_integration_message_id
+  }
+
+  /// Returns the previous diff message id, or the previous integration message id.
+  pub fn diff_message_id(&self) -> &MessageId {
+    if self.previous_diff_message_id.is_null() {
+      &self.previous_integration_message_id
+    } else {
+      &self.previous_diff_message_id
+    }
+  }
```
Nit: it might be better to return `Option<&MessageId>` and move the logic where it defaults to the `previous_integration_message_id` to where it's used?

PhilippGackstatter replied:
Good suggestion! I would still use a bare `MessageId`, though, since that has a `null` variant in itself. With `Option<MessageId>` you'd still have to technically check whether it is null. Since it's only used internally, that seems fine.

cycraig replied:
You're absolutely right, I forgot that `MessageId` needs to be checked for `null` regardless so returning `&MessageId` is absolutely fine. Might need to note in the doc comment here and in `previous_integration_message_id` that it could be `null`.

cycraig reviewed (changes requested):
Looks good overall. I like the simplification of the `Account` and the refactor seems to be on a good trajectory.

Minor changes from my side. I didn't notice any glaring issues with setting the message ids after a first pass but will take another look later.

PhilippGackstatter commented on identity-iota/src/did/doc/iota_verification_method.rs:
```diff
@@ -74,22 +75,22 @@ impl IotaVerificationMethod {
   }
 
   /// Creates a new [`IotaVerificationMethod`] from the given `did` and `keypair`.
-  pub fn from_did(did: IotaDID, keypair: &KeyPair, fragment: &str) -> Result<Self> {
+  pub fn from_did(did: IotaDID, key_type: KeyType, public_key: &PublicKey, fragment: &str) -> Result<Self> {
```
I think you're right about the reason being the Wasm bindings. So in `WasmVerificationMethod` I'll change it to:

```rust
pub fn from_did(did: &WasmDID, key_type: KeyType, public_key: String, fragment: String) -> ...
```

Regarding the default, I don't really like the `from_keypair` method, but doesn't it steer users in the right direction to produce a DID where the initial document is signed by the same key that produced the DID tag?

cycraig replied:
Users typically create documents through the `IotaDocument::new` constructor which derives the `DID` automatically. Users only seem to call `IotaVerificationMethod::new` when trying to add new methods to the document afterwards, which inevitably end up with the incorrect (derived) DID in the `id` field.

cycraig replied:
```
pub fn from_did(did: &WasmDID, key_type: KeyType, public_key: String, fragment: String) -> ...
```

This doesn't solve the problem of accidentally publishing one's `private_key` though which the Rust `PublicKey` type helps prevent. My suggestion would be to expose similar types in Wasm for the `KeyPair` instead of returning strings (and move/rename the string functions).

PhilippGackstatter replied:
Also, what I don't like about my change here, is that it's possible to provide a different `KeyType` than what the public key represents. We could either store the `KeyType` additionally into the `PublicKey` and `PrivateKey`, or preferably, create a thin wrapper `TypedPublicKey` and `TypedPrivateKey`, that would be the type of the parameter here.

That's probably overdoing... we can also note in the docs that the caller is responsible for providing the right types (it's the low-level API after all).

PhilippGackstatter replied:
> This doesn't solve the problem of accidentally publishing one's private_key in Wasm

I missed that point in your original post. That is definitely worth changing!

That seems a bit out of scope for this PR, so I'm thinking of reverting the Wasm changes for now and address this properly in a future PR, so the state refactor PR can land sooner.

PhilippGackstatter commented on identity-did/src/verification/method_scope.rs:
```diff
@@ -51,3 +53,15 @@ impl FromStr for MethodScope {
   }
 }
+
+impl From<MethodRelationship> for MethodScope {
+  fn from(relationship: MethodRelationship) -> Self {
+    match relationship {
+      MethodRelationship::Authentication => Self::Authentication,
+      MethodRelationship::AssertionMethod => Self::AssertionMethod,
+      MethodRelationship::KeyAgreement => Self::KeyAgreement,
+      MethodRelationship::CapabilityDelegation => Self::CapabilityDelegation,
+      MethodRelationship::CapabilityInvocation => Self::CapabilityInvocation,
+    }
+  }
+}
```
Ergonomics was the primary reason I copied the variants. I think we don't expect more relationships to be added any time soon, so the maintenance burden of keeping this up-to-date in two places is small.

```rust
// snippet
.create_method()
.scope(MethodScope::VerificationRelationship(MethodRelationship::AssertionMethod))
```

Nested seems acceptable, too. For Wasm in `insert_method`, as an example, we can also parse the string `"AssertionMethod"` into the nested variant, so ergonomics wouldn't be affected. Though, in the future, we might want to rethink whether to take a proper `WasmMethodScope` type there?

cycraig replied:
>  Though, in the future, we might want to rethink whether to take a proper WasmMethodScope type there?

Although I much prefer the nested version of `MethodScope`, since wasm_bindgen only supports flat enums we might need to flatten it manually for the Wasm bindings anyway.

I would be in favour of nesting it in Rust since that seems to make the most sense, and flattening it in Wasm when we implement `WasmMethodScope` (doesn't have to be in this PR).

E.g.
In Rust:
```rust
pub enum MethodScope {
  VerificationMethod,
  VerificationRelationship(MethodRelationship),
}
```
In Wasm bindings:
```rust
pub enum WasmMethodScope {
  VerificationMethod,
  Authentication,
  AssertionMethod,
  KeyAgreement,
  CapabilityDelegation,
  CapabilityInvocation,
}
```
with appropriate `From` implementations for conversions. 

One could also just wrap the `MethodScope` instead, which is how some other enums have been handled.
E.g.
```rust
pub struct WasmMethodScope(MethodScope);
```
Then expose the variants as functions.

PhilippGackstatter replied:
Nested it on the Rust side now. Wasm still works with strings for now. I think that's a good upgrade to do when getting the Wasm bindings ready for 1.0.

PhilippGackstatter commented on identity-did/src/verification/method_scope.rs:
```diff
@@ -6,6 +6,8 @@ use core::str::FromStr;
 use crate::error::Error;
 use crate::error::Result;
 
+use super::MethodRelationship;
```
No idea about import conventions, to be honest. I just rely on my IDE for that most of the time. I'm happy to change it!

PhilippGackstatter commented on identity-iota/src/did/doc/iota_document.rs:
```diff
@@ -1601,6 +1650,90 @@ mod tests {
+    // len is 1; the relationship was removed.
+    assert_eq!(document.as_document().verification_relationships().count(), 1);
+
+    // Removing it a second time does nothing
+    assert!(document
+      .detach_method_relationship(
+        document.did().to_url().join(fragment).unwrap(),
+        MethodRelationship::AssertionMethod,
+      )
+      .is_ok());
```
I think we had some agreement on moving the existence and duplication checks for `insert_method` and `remove_method` to the lower-level APIs. From that perspective, I suppose the relationship methods should do the same? If it's fine for now, then I would address this together with moving the method checks in another PR, as this is a slightly larger change, affecting error types, the `IotaDocument` as well as the account updates.

In the meantime, I'll change `detach_method_relationship` to return `Result<bool>` for at least some consistency.

cycraig replied:
> In the meantime, I'll change detach_method_relationship to return Result<bool> for at least some consistency.
>
> Edit: Can't change the return type, because OrderedSet::remove doesn't return bool.

I think we can check for existence in `OrderedSet::remove` first, much like `OrderedSet::append`:

```rust
  /// Adds a new value to the end of the `OrderedSet`; returns `true` if the
  /// value was successfully added.
  pub fn append(&mut self, item: T) -> bool
  where
    T: KeyComparable,
  {
    if self.contains(&item) {
      false
    } else {
      self.0.push(item);
      true
    }
  }
```

PhilippGackstatter commented on identity-iota/src/did/doc/iota_document.rs:
```diff
@@ -380,6 +385,49 @@ impl IotaDocument {
+    // Ensure the method exists
+    self
+      .document
+      .resolve_method(did_url.url())
+      .ok_or(identity_did::Error::QueryMethodNotFound)?;
+
+    let core_did_url: CoreDIDUrl = CoreDIDUrl::from(did_url);
+    let method_ref = MethodRef::Refer(core_did_url);
+
+    let result = match relationship {
+      MethodRelationship::Authentication => self.document.authentication_mut().append(method_ref),
+      MethodRelationship::AssertionMethod => self.document.assertion_method_mut().append(method_ref),
+      MethodRelationship::KeyAgreement => self.document.key_agreement_mut().append(method_ref),
+      MethodRelationship::CapabilityDelegation => self.document.capability_delegation_mut().append(method_ref),
+      MethodRelationship::CapabilityInvocation => self.document.capability_invocation_mut().append(method_ref),
+    };
+
+    Ok(result)
+  }
+
+  // Detaches the given relationship from the given method, if the method exists.
+  pub fn detach_method_relationship(&mut self, did_url: IotaDIDUrl, relationship: MethodRelationship) -> Result<()> {
+    // Ensure the method exists
+    self
+      .document
+      .resolve_method(did_url.url())
+      .ok_or(identity_did::Error::QueryMethodNotFound)?;
+
+    let core_did_url: CoreDIDUrl = CoreDIDUrl::from(did_url);
+
+    match relationship {
+      MethodRelationship::Authentication => self.document.authentication_mut().remove(&core_did_url),
+      MethodRelationship::AssertionMethod => self.document.assertion_method_mut().remove(&core_did_url),
+      MethodRelationship::KeyAgreement => self.document.key_agreement_mut().remove(&core_did_url),
+      MethodRelationship::CapabilityDelegation => self.document.capability_delegation_mut().remove(&core_did_url),
+      MethodRelationship::CapabilityInvocation => self.document.capability_invocation_mut().remove(&core_did_url),
+    }
+
+    Ok(())
+  }
```
Really good point, will do!

PhilippGackstatter commented on identity-account/src/account/account.rs:
```diff
@@ -193,58 +206,45 @@ impl Account {
-    Ok(())
+  #[doc(hidden)]
+  pub async fn load_state(&self) -> Result<IdentityState> {
+    // TODO: An account always holds a valid identity,
+    // so if None is returned, that's a broken invariant.
+    // This should be mapped to a fatal error in the future.
```
Definitely! This comment only applies to the mapping of `None` to `IdentityNotFound`, which can never occur, unless there is a logic bug in the account. The `Result` from the underlying storage is still returned, i.e. the `?` in that line. This is about not putting actual fatal errors ("stronghold file disappeared") together with errors that are practically guaranteed not to happen ("identity not found").

cycraig replied:
Thanks for the explanation!

PhilippGackstatter commented on identity-iota/src/did/doc/iota_document.rs:
```diff
@@ -380,6 +385,49 @@ impl IotaDocument {
+  }
+
+  // Detaches the given relationship from the given method, if the method exists.
+  pub fn detach_method_relationship(&mut self, did_url: IotaDIDUrl, relationship: MethodRelationship) -> Result<()> {
+    // Ensure the method exists
+    self
+      .document
+      .resolve_method(did_url.url())
+      .ok_or(identity_did::Error::QueryMethodNotFound)?;
+
+    let core_did_url: CoreDIDUrl = CoreDIDUrl::from(did_url);
```
The query is a good suggestion, makes it more generic. Unfortunately, we have to introduce a clone, because
of `cannot borrow *self as mutable because it is also borrowed as immutable` (mutably borrowed when mutating the relationships. But I changed it to use the query.

olivereanderson commented on identity-account/src/account/account.rs:
```diff
@@ -119,25 +88,58 @@ impl Account {
+    &self.state
+  }
+
+  pub(crate) fn state_mut_unchecked(&mut self) -> &mut IdentityState {
+    &mut self.state
+  }
```
I don't think we should mark this as `unsafe` as this function does not violate memory safety. Unsafe operations [are defined](https://doc.rust-lang.org/reference/unsafety.html) to be :

> Unsafe operations are those that can potentially violate the memory-safety guarantees of Rust's static semantics.

cycraig replied:
Perhaps, but we already have functions like `IotaDocument::as_document_mut` that are marked as `unsafe` because they can break invariants and essentially corrupt data (but not memory) due to unchecked changes, much like `state_mut_unchecked`. 

I would still prefer removing `state_mut_unchecked`.

PhilippGackstatter replied:
I will remove it.

I also would like us to re-evaluate what methods we mark `unsafe`. I agree it should be restricted to those that can violate memory-safety. For others, `_unchecked` and warning documentation seems good enough.

olivereanderson replied:
>we already have functions like IotaDocument::as_document_mut that are marked as unsafe because they can break invariants and essentially corrupt data (but not memory) due to unchecked changes, much like state_mut_unchecked.

Then we should consider changing those methods. The `unsafe` keyword is not meant to be used in this way.

cycraig replied:
`IotaDocument::as_document_mut` should be marked as `unsafe` because it allows the user to make changes that would violate other `unsafe` casts in `IotaDocument`. See all the comments in `validate_core_document`.

PhilippGackstatter commented on identity-account/src/identity/chain_state.rs:
```diff
@@ -0,0 +1,63 @@
+  }
+}
+
+impl Default for ChainState {
+  fn default() -> Self {
+    Self::new()
+  }
+}
```
Thanks for the hint, I had to double-check as well, but it requires all fields to implement `Default`, which is not the case for `MessageId`. Weirdly, given that it has `MessageId::null()` - should open a PR.

cycraig replied:
My bad, I'm used to clippy yelling at me to implement/derive `Default` whenever there is a plain `new()` constructor.

PhilippGackstatter commented on identity-account/src/account/account.rs:
```diff
@@ -255,166 +255,152 @@ impl Account {
+    if self.chain_state().is_new_identity() {
+      let method: &IotaVerificationMethod = new_state.document().default_signing_method()?;
+      let location: KeyLocation = new_state.method_location(
+        method.key_type(),
+        method
+          .id()
+          .fragment()
+          .expect("verification method did not have a fragment")
+          .to_owned(),
+      )?;
 
       // Sign the DID Document with the current capability invocation method
       new_state
-        .sign_data(self.did(), self.storage(), location, document)
+        .sign_data(self.did(), self.storage(), &location, document)
         .await?;
     } else {
-      let method: &TinyMethod = old_state.capability_invocation()?;
-      let location: &KeyLocation = method.location();
+      let method: &IotaVerificationMethod = old_state.document().default_signing_method()?;
+      let location: KeyLocation = new_state.method_location(
+        method.key_type(),
+        method
+          .id()
+          .fragment()
+          .expect("verification method did not have a fragment")
+          .to_owned(),
+      )?;
```
Bubbling up comes at little cost, but I'd have to add that variant to the error enum first, which adds to the number of errors a developer has to consider. It pollutes the error enum. No one would actually expect this error when calling `publish_changes`, because the user isn't providing any fragment here. If they had provided an erroneous fragment at some earlier point, they would've expected that method to error.

There seems to be a divide in the Rust community in general, where people say "panicking when encountering a broken invariant is fine", while others avoid it at all cost. I lean towards the former statement, but I'm also okay with us setting the guideline as "avoid panicking" for identity.rs. In that case, I would like us to consider a `FatalError` catch-all variant for these errors, so we bubble them up and give the user a chance to do some clean up or emergency handling. But returning `FatalError("method missing fragment")` seems totally fine here. In fact, Oliver is working on something like that in #487, so we can also take the discussion there, once it's ready.

For now, I'll add the variant and a `TODO` so we can reconsider in a future error-focused PR.

olivereanderson replied:
>No one would actually expect this error when calling publish_changes, because the user isn't providing any fragment here. If they had provided an erroneous fragment at some earlier point, they would've expected that method to error.
When this is the case it is indeed dangerous to bundle `MethodMissingFragment` together with failures we are able to explain as it is then not clearly conveyed that the program has now entered an unexpected state. 

In general there is a lot to be said about panicking early. It can prevent a corrupted program from producing disastrous side effects and it can also help us catch bugs early on in cases where our assumptions and/or program logic is wrong.  We should also keep in mind that panics in Rust is actually quite graceful in the sense that all `Drop` implementations get to run and it doesn't necessarily affect other threads, not to mention the generally helpful error message it yields. 

Having said that, one can of course question whether our library gets to decide when to panic. An alternative would be to follow the `sled` post mentioned in [this discussion](https://github.com/iotaledger/identity.rs/discussions/460) and return something like `Result<Result<(),ExpectedError>,FatalError>` which the caller may choose to either bubble up or panic from.  

Regardless of what is said above any function that can panic must document this in its documentation, and since one often does not bother documenting private functions, they probably should return the nested `Result` with an outer `FatalError`.

PhilippGackstatter commented on identity-account/src/account/account.rs:
```diff
@@ -193,58 +206,45 @@ impl Account {
+    let method: &IotaVerificationMethod = state.document().resolve_method(fragment).ok_or(Error::MethodNotFound)?;
 
-    state.sign_data(self.did(), self.storage(), location, target).await?;
+    let location: KeyLocation = state.method_location(method.key_type(), fragment.to_owned())?;
```
Good catch. I believe with the upcoming key refactor this should be addressed adequately, since the key location will be derived from the document consisting of fragment + a hash of the `MethodData`, which should only point methods with the same fragment and different DIDs to the same location, if the key material is the same, in which case that is the correct behavior.

PhilippGackstatter commented on identity-account/src/account/account.rs:
```diff
@@ -55,37 +46,15 @@ pub struct Account {
   storage: Arc<dyn Storage>,
   client_map: Arc<ClientMap>,
   actions: AtomicUsize,
-  did: IotaDID,
+  chain_state: ChainState,
+  state: IdentityState,
```
I forgot to add this logic back into `process_update`, but technically we still have the `Batch` save in `Account::save`, so I'll re-add it. This will be decided with #432 anyway. Thanks!

PhilippGackstatter commented:
> Can't we store the altered document instead of the diffs?

That should also work, good idea!

> I didn't notice any glaring issues with setting the message ids

Main reason I'm confused is that I didn't find where the previous message ids were set in the current code, specifically for integration updates.

I addressed most comments, except for the `attach_method_relationship` discussion (which I'm thinking of pushing to a later PR, when also moving more checks to the lower level for `remove_method` and `insert_method`) and the publish type with the `UPDATE_METHOD_TYPES` which I still want to add, to reduce the number of cases that require integration updates. Will work on that on Monday (presumably).

olivereanderson reviewed (approved):
I like the smaller crate level `Error` introduced in this PR. In general I would prefer to always return a named error instead of using an alias for the entire Result which seems to be mostly the case in this crate, but that is something we can think about when we get round to refactoring more of our error handling.

cycraig commented on identity-account/src/account/account.rs:
```diff
@@ -255,166 +252,151 @@ impl Account {
-    // Bail early if there are no new events
-    if events.is_empty() {
-      return Ok(Vec::new());
+    // If there was no previous diff message, use the previous int message.
+    if diff_id.is_null() {
+      diff_id = self.chain_state.last_integration_message_id();
```
Maybe we should return an error after this if the `diff_id` is still `null`? It would mean something went wrong with the chain state but it would avoid publishing an orphaned diff update with no chain.

PhilippGackstatter replied:
We can check, but the `diff_id` can only be null here when storing the chain state after the last update failed or the chain state got corrupted in some other way, in which case I would like to return a fatal error. When it happens, it's not the user's fault, so a fatal error seems appropriate. I would like to avoid adding another edge case error type, as that will just make life harder for the error refactoring, and instead leave a TODO there.

cycraig replied:
I'm not sure about the error variants but I do think we need something to indicate that the diff update will be invalid. A TODO should be fine for now.

cycraig commented on identity-account/src/account/account.rs:
```diff
@@ -255,166 +252,151 @@ impl Account {
-      .try_fold(initial, Self::fold_snapshot)
-      .await
-  }
+    let mut diff_id: &MessageId = self.chain_state().last_diff_message_id();
```
Nit: maybe `previous_message_id` or `diff_previous_message_id` would be more descriptive? Completely optional.

cycraig reviewed (approved):
Looks good to me, thanks for the changes!

Minor comments. I could have missed something between all the changes to be honest.

PhilippGackstatter commented on identity-iota/src/tangle/publish.rs:
```diff
@@ -0,0 +1,264 @@
+        MethodRef::Embed(method) => Some(method),
+        MethodRef::Refer(did_url) => old_doc.as_document().resolve_method(did_url),
+      })
+      .filter(|method| {
+        if let Some(method) = method {
+          UPDATE_METHOD_TYPES.contains(&method.key_type())
+        } else {
+          true
+        }
+      })
```
Oh, I thought this is what you meant in [this comment](https://github.com/iotaledger/identity.rs/pull/462#discussion_r747741254). I think if we decide that `MerkleKeyCollection`s will never be able to sign document updates, then it's fine. Any new signature type we introduce, we have to make a decision for whether to add or leave it out of `UPDATE_METHOD_TYPES`.

cycraig replied:
It was ^^ I was just questioning my own decision since the ramifications for future breaking changes didn't occur to me at the time.

Absolutely fine leaving it as-is for now to be clear.
